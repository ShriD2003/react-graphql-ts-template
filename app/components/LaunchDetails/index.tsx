import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { If, T, For } from '@components';
import isEmpty from 'lodash-es/isEmpty';
import { colors, media } from '@app/themes';
import { LaunchDetails as LaunchDetailsType } from '@app/containers/LaunchDetails/types';
import { Card, Image, Skeleton } from 'antd';
import placeholderImage from '@images/undraw_to_the_stars_re_wq2x.svg';

const LaunchDetailsCard = styled(Card)`
  && {
    .ant-card-body {
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      color: ${(props) => props.color};

      ${media.greaterThan('tablet')`
        flex-direction: row;
      `}
    }
    margin: 1.5rem;
    background-color: ${colors.cardBg};
  }
`;

const CustomImage = styled(Image)`
  && {
    ${media.between('tablet', 'desktop')`
    .ant-image-img{
      width: 80%;
      margin: 0;
    }
    `}
    ${media.greaterThan('desktop')`
      height: 692px;
      width: 547px;
      margin: 2rem;
    `}
  }
`;

const DetailsCard = styled.div`
  && {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 1rem;
    margin: 3rem 0;
    ${media.greaterThan('desktop')`
      width: 50%;
      /* margin: 13% 2%; */
   `}
  }
`;

const RocketBox = styled.div`
  && {
    display: flex;
    gap: 1rem;
    margin-left: 1.5rem;
  }
`;

const CustomT = styled(T)`
  && {
    display: block;
    font-weight: thin;
    opacity: 0.7;
  }
`;

const LaunchLabel = styled.b`
  text-transform: uppercase;
  display: block;
  font-size: small;
  color: ${colors.primary};
  opacity: 1;
  font-weight: bolder;
`;

const ShipContainer = styled.div`
  margin-left: 1.5rem;
  display: grid;
  grid-template-columns: max-content auto;
  align-items: center;
  column-gap: 1rem;
  row-gap: 1rem;
`;

function labelRenderer(chunks: string) {
  return <LaunchLabel>{chunks}</LaunchLabel>;
}

interface LaunchDetailsProps extends LaunchDetailsType {
  loading: boolean;
}

function LaunchDetails({ missionName, links, details, rocket, ships, loading }: LaunchDetailsProps) {
  return (
    <LaunchDetailsCard data-testid="launch-details">
      <Skeleton loading={loading} active>
        <If
          condition={!isEmpty(links?.flickrImages)}
          otherwise={<CustomImage preview={false} src={placeholderImage} />}
        >
          <CustomImage preview={false} src={links?.flickrImages![0]} />
        </If>
      </Skeleton>
      <Skeleton loading={loading} active>
        <DetailsCard>
          <If condition={!isEmpty(missionName)}>
            <CustomT marginBottom={0.5} data-testid="mission-name" type="heading" text={missionName} />
          </If>

          <If condition={!isEmpty(details)}>
            <CustomT data-testid="details" type="standard" id="details" values={{ details, b: labelRenderer }} />
          </If>
          <If condition={!isEmpty(rocket)}>
            <LaunchLabel>Rocket </LaunchLabel>
            <RocketBox>
              <If condition={!isEmpty(rocket?.rocketName)}>
                <CustomT
                  data-testid="rocket-name"
                  type="standard"
                  id="name_label"
                  values={{ name: rocket?.rocketName, b: labelRenderer }}
                />
              </If>
              <If condition={!isEmpty(rocket?.rocketType)}>
                <CustomT
                  data-testid="rocket-type"
                  type="standard"
                  id="type_label"
                  values={{ type: rocket?.rocketType, b: labelRenderer }}
                />
              </If>
            </RocketBox>
          </If>
          <If condition={!isEmpty(ships)}>
            <LaunchLabel>Ships </LaunchLabel>
            <ShipContainer>
              <CustomT marginBottom={-0.5} id="name_label" values={{ name: '', b: labelRenderer }} />
              <CustomT marginBottom={-0.5} id="type_label" values={{ type: '', b: labelRenderer }} />
              <For
                noParent
                of={ships}
                renderItem={(ship) => (
                  <>
                    <If condition={!isEmpty(ship.name)}>
                      <CustomT data-testid="ship-name" type="standard" text={ship.name} />
                    </If>
                    <If condition={!isEmpty(ship.type)}>
                      <CustomT data-testid="ship-type" type="standard" text={ship.type} />
                    </If>
                  </>
                )}
              />
            </ShipContainer>
          </If>
        </DetailsCard>
      </Skeleton>
    </LaunchDetailsCard>
  );
}

LaunchDetails.propTypes = {
  id: PropTypes.string,
  loading: PropTypes.bool,
  missionName: PropTypes.string,
  launchDateUtc: PropTypes.string,
  links: PropTypes.shape({
    flickrImages: PropTypes.arrayOf(PropTypes.string)
  }),
  details: PropTypes.string,
  rocket: PropTypes.shape({
    rocketName: PropTypes.string,
    rocketType: PropTypes.string
  }),
  ships: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      type: PropTypes.string
    })
  )
};

export default LaunchDetails;
