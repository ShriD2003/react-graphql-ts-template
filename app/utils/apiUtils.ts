import { ApisauceInstance, create } from 'apisauce';
import snakeCase from 'lodash/snakeCase';
import camelCase from 'lodash/camelCase';
import { mapKeysDeep } from './index';

const apiClients: Record<string, ApisauceInstance> = {};

export const getApiClient = (type = 'github') => apiClients[type];

export const generateApiClient = (type = 'github') => {
  switch (type) {
    case 'github':
      apiClients[type] = createApiClientWithTransForm(process.env.GITHUB_URL!);
      return apiClients[type];
    default:
      apiClients.default = createApiClientWithTransForm(process.env.GITHUB_URL!);
      return apiClients.default;
  }
};

export const createApiClientWithTransForm = (baseURL: string) => {
  const api = create({
    baseURL,
    headers: { 'Content-Type': 'application/json' }
  });
  api.addResponseTransform((response) => {
    const { ok, data } = response;
    if (ok && data) {
      response.data = mapKeysDeep(data, (keys: string) => camelCase(keys));
    }
    return response;
  });

  api.addRequestTransform((request) => {
    const { data } = request;
    if (data) {
      request.data = mapKeysDeep(data, (keys: string) => snakeCase(keys));
    }
    return request;
  });
  return api;
};
