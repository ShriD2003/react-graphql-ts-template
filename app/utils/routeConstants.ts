export type RouteConstant = {
  route: string;
  exact?: boolean;
  isProtected?: boolean;
  props?: object;
};

const routeConstants: Record<string, RouteConstant> = {
  home: {
    route: '/',
    exact: true
  },
  launch: {
    route: '/:launchId',
    exact: true
  }
};

export type RouteKeys = keyof typeof routeConstants;

export default routeConstants;
