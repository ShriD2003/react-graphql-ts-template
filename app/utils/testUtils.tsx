import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
