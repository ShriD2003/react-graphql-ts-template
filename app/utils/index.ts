import find from 'lodash-es/find';
import get from 'lodash-es/get';
import history from './history';
import routeConstants from './routeConstants';

export const getCurrentRouteDetails = (location: Partial<Location>) => {
  if (!get(location, 'pathname')) {
    return null;
  }
  const route = find(
    Object.keys(routeConstants),
    (key) => routeConstants[key].route === location.pathname || `${routeConstants[key].route}/` === location.pathname
  );
  if (route) {
    return routeConstants[route];
  }
  return null;
};
export const mapKeysDeep = <T>(obj: T, fn: (key: string) => string): T =>
  Array.isArray(obj)
    ? obj.map((val) => mapKeysDeep(val, fn))
    : typeof obj === 'object'
    ? Object.keys(obj).reduce((acc, current) => {
        const key = fn(current);
        const val = obj[current as keyof T];
        acc[key] = val !== null && typeof val === 'object' ? mapKeysDeep(val, fn) : val;
        return acc;
      }, {} as any)
    : obj;

export const isLocal = () => {
  try {
    if (process.env.IS_LOCAL) {
      const local = JSON.parse(process.env.IS_LOCAL);
      return typeof local === 'boolean' && local;
    }
  } catch {
    // continue regardless of error
  }
  return false;
};

export const setQueryParam = (param: string, value: string | number, op: 'set' | 'delete' = 'set') => {
  const urlParams = new URLSearchParams(history.location.search);
  urlParams[op](param, value as string);
  history.push({ search: urlParams.toString() });
};
