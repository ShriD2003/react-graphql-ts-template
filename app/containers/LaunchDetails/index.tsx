/**
 *
 * LaunchDetails
 *
 */

import React, { memo, useEffect } from 'react';
import LaunchDetailsComponent from '@app/components/LaunchDetails';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { AnyAction, compose } from 'redux';
import { injectSaga } from 'redux-injectors';
import { selectLaunch, selectLaunchError, selectLoading } from './selectors';
import saga, { LaunchDetails as LaunchDetailsType } from './saga';
import { requestGetLaunch } from './reducer';
import { useParams } from 'react-router-dom';
import { Skeleton } from 'antd';

export interface LaunchDetailsProps {
  launch: LaunchDetailsType | null;
  launchError?: string;
  loading: boolean;
  dispatchLaunch: (launchId: string) => AnyAction;
}

export function LaunchDetails({ launch, launchError, loading, dispatchLaunch }: LaunchDetailsProps) {
  const params = useParams<{ launchId?: string }>();

  useEffect(() => {
    if (params.launchId) {
      dispatchLaunch(params.launchId);
    }
  }, [params]);

  return (
    <div>
      <Helmet>
        <title>Launch Details</title>
        <meta name="description" content="Description of LaunchDetails" />
      </Helmet>
      <Skeleton loading={loading} active>
        {launch && <LaunchDetailsComponent {...launch} />}
        <p>{launchError}</p>
      </Skeleton>
    </div>
  );
}

LaunchDetails.propTypes = {
  launch: PropTypes.object,
  launchError: PropTypes.string,
  loading: PropTypes.bool.isRequired,
  dispatchLaunch: PropTypes.func.isRequired
};

const mapStateToProps = createStructuredSelector({
  launch: selectLaunch(),
  launchError: selectLaunchError(),
  loading: selectLoading()
});

export function mapDispatchToProps(dispatch: Function) {
  return {
    dispatchLaunch: (launchId: string) => dispatch(requestGetLaunch(launchId))
  };
}

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect, memo, injectSaga({ key: 'launchDetails', saga }))(LaunchDetails);

export const LaunchDetailsTest = LaunchDetails;
