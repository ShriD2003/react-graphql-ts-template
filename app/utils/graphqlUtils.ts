import { Launch } from '@app/containers/HomeContainer';
import ApolloClient, { DocumentNode, InMemoryCache, QueryOptions } from 'apollo-boost';

export const client = new ApolloClient({
  uri: 'https://api.spacex.land/graphql',
  cache: new InMemoryCache()
});

interface ResponseData {
  data: Launch | null;
  ok: boolean | null;
  error: Object | null;
}

export const getQueryResponse = (query: DocumentNode, variables?: QueryOptions['variables']) => {
  const responseData: ResponseData = {
    data: null,
    error: null,
    ok: false
  };
  return client
    .query<ResponseData>({ query: query, variables })
    .then((res) => {
      if (res.errors) {
        return { ...responseData, error: res.errors };
      } else {
        return { ...responseData, data: res, ok: true };
      }
    })
    .catch((err) => {
      return { ...responseData, error: err };
    });
};
