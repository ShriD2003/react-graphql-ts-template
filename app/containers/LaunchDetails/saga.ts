import { getQueryResponse, GqlQueryReponse } from '@app/utils/graphqlUtils';
import { AnyAction } from 'redux';
import { call, takeLatest, put } from 'redux-saga/effects';
import { GET_LAUNCH } from './queries';
import { requestGetLaunch, successGetLaunch, failureGetLaunch } from './reducer';
// Individual exports for testing

export interface LaunchDetails {
  id: string;
  missionName: string;
  details: string;
  rocket: {
    rocketName: string;
    rocketType: string;
  };
  ships: {
    name: string;
    type: string;
  }[];
  links: {
    flickrImages?: string[];
  };
}

type LaunchResponse = GqlQueryReponse<LaunchDetails>;

export function* getLaunch(action: AnyAction): Generator<any, any, LaunchResponse> {
  const response = yield call(getQueryResponse, GET_LAUNCH, { launchId: action.payload });

  const { ok, data, error } = response;

  if (ok) {
    yield put(successGetLaunch(data));
  } else {
    yield put(failureGetLaunch(error));
  }
}

export default function* launchDetailsSaga() {
  yield takeLatest(requestGetLaunch.toString(), getLaunch);
}
