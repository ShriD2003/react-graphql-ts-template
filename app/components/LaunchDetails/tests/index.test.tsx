import React from 'react';
import LaunchDetails from '../index';
import { renderProvider } from '@app/utils/testUtils';

describe('<LaundDetails> tests', () => {
  const launchDetails = {
    id: '1',
    missionName: 'CRS-21',
    links: {
      flickrImages: ['https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg']
    },
    details: "SpaceX's 21st ISS resupply mission.",
    rocket: {
      rocketName: 'Falcon 9',
      rocketType: 'FT'
    },
    ships: [
      {
        name: 'Ship 1',
        type: 'Type 1'
      }
    ]
  };
  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render the mission name if it is available', () => {
    const { getByTestId } = renderProvider(<LaunchDetails {...launchDetails} />);
    expect(getByTestId('mission-name')).toBeInTheDocument();
  });
});
