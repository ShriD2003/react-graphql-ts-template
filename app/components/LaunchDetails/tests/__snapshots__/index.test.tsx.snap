// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaundDetails> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered LaunchDetails__LaunchDetailsCard-ytnak2-0 diIhJW"
      data-testid="launch-details"
    >
      <div
        class="ant-card-body"
      >
        <img
          class="LaunchDetails__CustomImage-ytnak2-1 gjpxoD"
          src="https://farm9.staticflickr.com/8617/16789019815_f99a165dc5_o.jpg"
        />
        <div
          class="LaunchDetails__DetailsCard-ytnak2-2 fSBLBB"
        >
          <p
            class="T__StyledText-gcrjip-0 kPLUDf LaunchDetails__CustomT-ytnak2-4 OzFVO"
            data-testid="mission-name"
          >
            CRS-21
          </p>
          <p
            class="T__StyledText-gcrjip-0 cSxVIW LaunchDetails__CustomT-ytnak2-4 OzFVO"
            data-testid="details"
          >
            <b
              class="LaunchDetails__LaunchLabel-ytnak2-5 cMgOxy"
            >
              Details
            </b>
             SpaceX's 21st ISS resupply mission.
          </p>
          <p
            class="T__StyledText-gcrjip-0 gWdzqB LaunchDetails__LaunchLabelT-ytnak2-6 cuMxwl"
            data-testid="t"
          >
            Rocket 
          </p>
          <div
            class="LaunchDetails__RocketBox-ytnak2-3 kQjTjv"
          >
            <p
              class="T__StyledText-gcrjip-0 cSxVIW LaunchDetails__CustomT-ytnak2-4 OzFVO"
              data-testid="rocket-name"
            >
              <b
                class="LaunchDetails__LaunchLabel-ytnak2-5 cMgOxy"
              >
                Name
              </b>
               Falcon 9
            </p>
            <p
              class="T__StyledText-gcrjip-0 cSxVIW LaunchDetails__CustomT-ytnak2-4 OzFVO"
              data-testid="rocket-type"
            >
              <b
                class="LaunchDetails__LaunchLabel-ytnak2-5 cMgOxy"
              >
                Type
              </b>
               FT
            </p>
          </div>
          <p
            class="T__StyledText-gcrjip-0 gWdzqB LaunchDetails__LaunchLabelT-ytnak2-6 cuMxwl"
            data-testid="t"
          >
            Ships 
          </p>
          <div
            class="LaunchDetails__ShipContainer-ytnak2-7 fAKuKZ"
          >
            <p
              class="T__StyledText-gcrjip-0 cSxVIW LaunchDetails__CustomT-ytnak2-4 OzFVO"
              data-testid="ship-name"
            >
              <b
                class="LaunchDetails__LaunchLabel-ytnak2-5 cMgOxy"
              >
                Name
              </b>
               Ship 1
            </p>
            <p
              class="T__StyledText-gcrjip-0 cSxVIW LaunchDetails__CustomT-ytnak2-4 OzFVO"
              data-testid="ship-type"
            >
              <b
                class="LaunchDetails__LaunchLabel-ytnak2-5 cMgOxy"
              >
                Type
              </b>
               Type 1
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
